In [1]:
import os

# Show the directory where this notebook is currently running
print("Current working directory:")
print(os.getcwd())

Current working directory:
/home/kusuma-rani-sutar/tourism-analytics/notebooks


In [2]:
# Show all files inside the raw data folder
print("Files in data/raw:")
print(os.listdir("../data/raw"))

Files in data/raw:
['Country.xlsx', 'Mode.xlsx', 'Updated_Item.xlsx', 'User.xlsx', 'Region.xlsx', 'Type.xlsx', 'Transaction.xlsx', 'Item.xlsx', 'City.xlsx', 'Continent.xlsx']


In [4]:
import pandas as pd

# Load each Excel table from the raw data folder
country = pd.read_excel("../data/raw/Country.xlsx")
mode = pd.read_excel("../data/raw/Mode.xlsx")
item_updated = pd.read_excel("../data/raw/Updated_Item.xlsx")
user = pd.read_excel("../data/raw/User.xlsx")
region = pd.read_excel("../data/raw/Region.xlsx")
type_ = pd.read_excel("../data/raw/Type.xlsx")
transaction = pd.read_excel("../data/raw/Transaction.xlsx")
item = pd.read_excel("../data/raw/Item.xlsx")
city = pd.read_excel("../data/raw/City.xlsx")
continent = pd.read_excel("../data/raw/Continent.xlsx")

print("All tables loaded successfully!")

All tables loaded successfully!


In [5]:
# Show the number of rows and columns in every table
tables = {
    "Country": country,
    "Mode": mode,
    "Updated_Item": item_updated,
    "User": user,
    "Region": region,
    "Type": type_,
    "Transaction": transaction,
    "Item": item,
    "City": city,
    "Continent": continent
}

for name, df in tables.items():
    print(f"{name:15} → {df.shape[0]:,} rows × {df.shape[1]} columns")

Country         → 165 rows × 3 columns
Mode            → 6 rows × 2 columns
Updated_Item    → 1,698 rows × 5 columns
User            → 33,530 rows × 5 columns
Region          → 22 rows × 3 columns
Type            → 17 rows × 2 columns
Transaction     → 52,930 rows × 7 columns
Item            → 30 rows × 5 columns
City            → 9,143 rows × 3 columns
Continent       → 6 rows × 2 columns


In [6]:
# Display the structure of every table
# This helps us understand column names, data types,
# and whether there are missing values.

for name, df in tables.items():
    print("=" * 70)
    print(f"TABLE: {name}")
    print("=" * 70)
    
    df.info()
    
    print()

TABLE: Country
<class 'pandas.DataFrame'>
RangeIndex: 165 entries, 0 to 164
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   CountryId  165 non-null    int64
 1   Country    165 non-null    str  
 2   RegionId   165 non-null    int64
dtypes: int64(2), str(1)
memory usage: 5.4 KB

TABLE: Mode
<class 'pandas.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   VisitModeId  6 non-null      int64
 1   VisitMode    6 non-null      str  
dtypes: int64(1), str(1)
memory usage: 261.0 bytes

TABLE: Updated_Item
<class 'pandas.DataFrame'>
RangeIndex: 1698 entries, 0 to 1697
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   AttractionId       1698 non-null   int64 
 1   AttractionCityId   1698 non-null   int64 
 2   AttractionTypeId   1698 non-null  

####  Checking Missing values

In [7]:
# Check missing values in every table

for name, df in tables.items():
    print("=" * 70)
    print(f"TABLE: {name}")
    print("=" * 70)
    
    print(df.isnull().sum())
    
    print()

TABLE: Country
CountryId    0
Country      0
RegionId     0
dtype: int64

TABLE: Mode
VisitModeId    0
VisitMode      0
dtype: int64

TABLE: Updated_Item
AttractionId         0
AttractionCityId     0
AttractionTypeId     0
Attraction           0
AttractionAddress    0
dtype: int64

TABLE: User
UserId         0
ContinentId    0
RegionId       0
CountryId      0
CityId         4
dtype: int64

TABLE: Region
Region         0
RegionId       0
ContinentId    0
dtype: int64

TABLE: Type
AttractionTypeId    0
AttractionType      0
dtype: int64

TABLE: Transaction
TransactionId    0
UserId           0
VisitYear        0
VisitMonth       0
VisitMode        0
AttractionId     0
Rating           0
dtype: int64

TABLE: Item
AttractionId         0
AttractionCityId     0
AttractionTypeId     0
Attraction           0
AttractionAddress    0
dtype: int64

TABLE: City
CityId       0
CityName     1
CountryId    0
dtype: int64

TABLE: Continent
ContinentId    0
Continent      0
dtype: int64



In [9]:
# Show the User records where CityId is missing

user[user["CityId"].isnull()]

,UserId,ContinentId,RegionId,CountryId,CityId
2279,7175,5,17,135,NaN
6027,17595,1,4,22,NaN
21303,56972,5,17,135,NaN
25494,67461,5,17,135,NaN


In [10]:
# Show the Transaction records associated with those users

missing_city_users = user[user["CityId"].isnull()]["UserId"]

transaction[transaction["UserId"].isin(missing_city_users)]

,TransactionId,UserId,VisitYear,VisitMonth,VisitMode,AttractionId,Rating
7134,10437,7175,2016,10,5,640,5
21190,33604,56972,2016,12,2,673,4
22283,35006,17595,2014,12,3,673,4
26249,42487,56972,2016,12,2,650,5
31689,67298,67461,2019,5,4,748,5
34600,71286,56972,2016,12,2,748,5
34799,71574,7175,2016,10,5,748,5
36610,74011,17595,2014,12,3,748,5


In [11]:
# Get the UserIds whose CityId is missing
missing_city_users = user.loc[user["CityId"].isnull(), "UserId"]

print("Users with missing CityId:")
print(missing_city_users.tolist())

print("\nNumber of transactions for each affected user:")

# Count transactions belonging to each affected user
print(
    transaction[transaction["UserId"].isin(missing_city_users)]
    .groupby("UserId")
    .size()
)

Users with missing CityId:
[7175, 17595, 56972, 67461]

Number of transactions for each affected user:
UserId
7175     2
17595    2
56972    3
67461    1
dtype: int64


In [12]:
# Find the city record with missing CityName
city[city["CityName"].isnull()]

,CityId,CityName,CountryId
6879,6879,NaN,151


In [13]:
# User records with missing CityId
missing_users = user[user["CityId"].isnull()]

# Show their user information
print("AFFECTED USERS")
display(missing_users)

# Show all transactions belonging to these users
print("\nAFFECTED TRANSACTIONS")
display(
    transaction[
        transaction["UserId"].isin(missing_users["UserId"])
    ].sort_values("UserId")
)

AFFECTED USERS


,UserId,ContinentId,RegionId,CountryId,CityId
2279,7175,5,17,135,NaN
6027,17595,1,4,22,NaN
21303,56972,5,17,135,NaN
25494,67461,5,17,135,NaN



AFFECTED TRANSACTIONS


,TransactionId,UserId,VisitYear,VisitMonth,VisitMode,AttractionId,Rating
7134,10437,7175,2016,10,5,640,5
34799,71574,7175,2016,10,5,748,5
36610,74011,17595,2014,12,3,748,5
22283,35006,17595,2014,12,3,673,4
26249,42487,56972,2016,12,2,650,5
21190,33604,56972,2016,12,2,673,4
34600,71286,56972,2016,12,2,748,5
31689,67298,67461,2019,5,4,748,5


In [14]:
# Find all information related to CityId 6879

print("City record:")
display(city[city["CityId"] == 6879])

print("\nItems/attractions associated with CityId 6879:")
display(item[item["AttractionCityId"] == 6879])

City record:


,CityId,CityName,CountryId
6879,6879,NaN,151



Items/attractions associated with CityId 6879:


,AttractionId,AttractionCityId,AttractionTypeId,Attraction,AttractionAddress


In [15]:
# Attractions visited by the users with missing CityId

affected_attraction_ids = transaction[
    transaction["UserId"].isin(missing_users["UserId"])
]["AttractionId"].unique()

print("Attraction IDs:")
print(affected_attraction_ids)

print("\nCorresponding attractions and cities:")

display(
    item[
        item["AttractionId"].isin(affected_attraction_ids)
    ][
        [
            "AttractionId",
            "AttractionCityId",
            "Attraction",
            "AttractionAddress"
        ]
    ].sort_values("AttractionId")
)

Attraction IDs:
[640 673 650 748]

Corresponding attractions and cities:


,AttractionId,AttractionCityId,Attraction,AttractionAddress
2,640,1,Sacred Monkey Forest Sanctuary,"Jl. Monkey Forest, Ubud 80571 Indonesia"
3,650,1,Sanur Beach,Sanur
4,673,1,Seminyak Beach,Seminyak
6,748,1,Tegalalang Rice Terrace,"Jalan Raya Ceking, Tegalalang 80517 Indonesia"


#### So we can confidently say that all four users visited attractions in the same city represented by AttractionCityId = 1.

In [16]:
# Find the city represented by CityId = 1

display(
    city[city["CityId"] == 1]
)

,CityId,CityName,CountryId
1,1,Douala,1


In [17]:
# Check the country for CityId = 1

city_1 = city[city["CityId"] == 1]

display(
    country[country["CountryId"].isin(city_1["CountryId"])]
)

,CountryId,Country,RegionId
1,1,Cameroon,1


The attractions visited by the four users with missing CityId all have:

AttractionCityId = 1

But:

CityId = 1 → Douala → Cameroon

while the attraction addresses say:

Sacred Monkey Forest Sanctuary → Ubud, Indonesia
Sanur Beach → Sanur, Indonesia
Seminyak Beach → Indonesia
Tegalalang Rice Terrace → Indonesia

That means the source data contains a mismatched ID relationship.

Therefore, we must NOT fill those four users' missing CityId with 1.

That would incorrectly turn users associated with CountryId 135/22 into Douala, Cameroon.


## Check for Duplicates

In [18]:
# ============================================================
# STEP 2: CHECK DUPLICATE ROWS IN EVERY TABLE
# ============================================================

for name, df in tables.items():
    print("=" * 70)
    print(f"TABLE: {name}")
    print("=" * 70)

    duplicate_count = df.duplicated().sum()

    print(f"Duplicate rows: {duplicate_count}")
    print()

TABLE: Country
Duplicate rows: 0

TABLE: Mode
Duplicate rows: 0

TABLE: Updated_Item
Duplicate rows: 0

TABLE: User
Duplicate rows: 0

TABLE: Region
Duplicate rows: 0

TABLE: Type
Duplicate rows: 0

TABLE: Transaction
Duplicate rows: 0

TABLE: Item
Duplicate rows: 0

TABLE: City
Duplicate rows: 0

TABLE: Continent
Duplicate rows: 0



### Check ID Consistency 

In [19]:
# ============================================================
# STEP 3: CHECK ID CONSISTENCY / REFERENTIAL INTEGRITY
# ============================================================

print("=" * 70)
print("1. TRANSACTION -> USER")
print("=" * 70)

transaction_user_ids = set(transaction["UserId"])
user_ids = set(user["UserId"])

missing_users = transaction_user_ids - user_ids

print("Unique UserIds in Transaction:", len(transaction_user_ids))
print("Unique UserIds in User:", len(user_ids))
print("Transaction UserIds not found in User:", len(missing_users))

if missing_users:
    print("Missing UserIds:", sorted(missing_users)[:20])


print("\n" + "=" * 70)
print("2. TRANSACTION -> ATTRACTION")
print("=" * 70)

transaction_attraction_ids = set(transaction["AttractionId"])

item_ids = set(item["AttractionId"])
updated_item_ids = set(item_updated["AttractionId"])

print("Unique AttractionIds in Transaction:", len(transaction_attraction_ids))
print("AttractionIds found in Item:", len(transaction_attraction_ids & item_ids))
print("AttractionIds found in Updated_Item:",
      len(transaction_attraction_ids & updated_item_ids))

print(
    "Transaction AttractionIds NOT found in Item:",
    len(transaction_attraction_ids - item_ids)
)

print(
    "Transaction AttractionIds NOT found in Updated_Item:",
    len(transaction_attraction_ids - updated_item_ids)
)


print("\n" + "=" * 70)
print("3. UPDATED_ITEM -> TYPE")
print("=" * 70)

print("Updated_Item AttractionTypeId dtype:",
      item_updated["AttractionTypeId"].dtype)

print("Type AttractionTypeId dtype:",
      type_["AttractionTypeId"].dtype)

print("\nSample Updated_Item AttractionTypeId values:")
print(item_updated["AttractionTypeId"].head(20).tolist())

print("\nType AttractionTypeId values:")
print(type_["AttractionTypeId"].tolist())


print("\n" + "=" * 70)
print("4. USER -> CITY")
print("=" * 70)

user_city_ids = set(user["CityId"].dropna())
city_ids = set(city["CityId"])

missing_city_references = user_city_ids - city_ids

print("Unique non-null CityIds in User:", len(user_city_ids))
print("Unique CityIds in City:", len(city_ids))
print("User CityIds not found in City:", len(missing_city_references))

if missing_city_references:
    print("Missing CityIds:", sorted(missing_city_references)[:20])


print("\n" + "=" * 70)
print("5. CITY -> COUNTRY")
print("=" * 70)

city_country_ids = set(city["CountryId"])
country_ids = set(country["CountryId"])

missing_country_references = city_country_ids - country_ids

print("City CountryIds not found in Country:",
      len(missing_country_references))

if missing_country_references:
    print("Missing CountryIds:",
          sorted(missing_country_references)[:20])


print("\n" + "=" * 70)
print("6. COUNTRY -> REGION")
print("=" * 70)

country_region_ids = set(country["RegionId"])
region_ids = set(region["RegionId"])

missing_region_references = country_region_ids - region_ids

print("Country RegionIds not found in Region:",
      len(missing_region_references))

if missing_region_references:
    print("Missing RegionIds:",
          sorted(missing_region_references)[:20])


print("\n" + "=" * 70)
print("7. REGION -> CONTINENT")
print("=" * 70)

region_continent_ids = set(region["ContinentId"])
continent_ids = set(continent["ContinentId"])

missing_continent_references = region_continent_ids - continent_ids

print("Region ContinentIds not found in Continent:",
      len(missing_continent_references))

if missing_continent_references:
    print("Missing ContinentIds:",
          sorted(missing_continent_references)[:20])

1. TRANSACTION -> USER
Unique UserIds in Transaction: 33530
Unique UserIds in User: 33530
Transaction UserIds not found in User: 0

2. TRANSACTION -> ATTRACTION
Unique AttractionIds in Transaction: 30
AttractionIds found in Item: 30
AttractionIds found in Updated_Item: 30
Transaction AttractionIds NOT found in Item: 0
Transaction AttractionIds NOT found in Updated_Item: 0

3. UPDATED_ITEM -> TYPE
Updated_Item AttractionTypeId dtype: object
Type AttractionTypeId dtype: int64

Sample Updated_Item AttractionTypeId values:
[13, 13, 63, 13, 13, 76, 72, 93, 76, 92, 13, 61, 93, 13, 64, 82, 72, 91, 84, 63]

Type AttractionTypeId values:
[2, 10, 13, 19, 34, 44, 45, 61, 63, 64, 72, 76, 82, 84, 91, 92, 93]

4. USER -> CITY
Unique non-null CityIds in User: 5545
Unique CityIds in City: 9143
User CityIds not found in City: 0

5. CITY -> COUNTRY
City CountryIds not found in Country: 0

6. COUNTRY -> REGION
Country RegionIds not found in Region: 0

7. REGION -> CONTINENT
Region ContinentIds not found 

What your results tell us

These relationships are clean:

Transaction → User             
Transaction → Attraction       
User → City                     (except 4 missing CityIds)
City → Country                 
Country → Region               
Region → Continent             

So we have no orphan foreign keys.

### Finish the ID/categorical investigation

In [20]:
# ============================================================
# STEP 3A: COMPARE ITEM AND UPDATED_ITEM
# ============================================================

print("=" * 70)
print("ITEM VS UPDATED_ITEM")
print("=" * 70)

print("Item shape:", item.shape)
print("Updated_Item shape:", item_updated.shape)

print("\nItem columns:")
print(item.columns.tolist())

print("\nUpdated_Item columns:")
print(item_updated.columns.tolist())


# Compare the attraction IDs used by Transaction
transaction_attraction_ids = set(transaction["AttractionId"])

print("\nAttraction IDs used by Transaction:", len(transaction_attraction_ids))

print(
    "Transaction IDs available in Item:",
    len(transaction_attraction_ids & set(item["AttractionId"]))
)

print(
    "Transaction IDs available in Updated_Item:",
    len(transaction_attraction_ids & set(item_updated["AttractionId"]))
)


# Compare the two tables for the 30 attractions actually used
item_used = (
    item[item["AttractionId"].isin(transaction_attraction_ids)]
    .sort_values("AttractionId")
    .reset_index(drop=True)
)

updated_used = (
    item_updated[item_updated["AttractionId"].isin(transaction_attraction_ids)]
    .sort_values("AttractionId")
    .reset_index(drop=True)
)

print("\nRows from Item used by Transaction:")
display(item_used)

print("\nRows from Updated_Item used by Transaction:")
display(updated_used)

ITEM VS UPDATED_ITEM
Item shape: (30, 5)
Updated_Item shape: (1698, 5)

Item columns:
['AttractionId', 'AttractionCityId', 'AttractionTypeId', 'Attraction', 'AttractionAddress']

Updated_Item columns:
['AttractionId', 'AttractionCityId', 'AttractionTypeId', 'Attraction', 'AttractionAddress']

Attraction IDs used by Transaction: 30
Transaction IDs available in Item: 30
Transaction IDs available in Updated_Item: 30

Rows from Item used by Transaction:


,AttractionId,AttractionCityId,AttractionTypeId,Attraction,AttractionAddress
0,369,1,13,Kuta Beach - Bali,Kuta
1,481,1,13,Nusa Dua Beach,"Semenanjung Nusa Dua, Nusa Dua 80517 Indonesia"
2,640,1,63,Sacred Monkey Forest Sanctuary,"Jl. Monkey Forest, Ubud 80571 Indonesia"
3,650,1,13,Sanur Beach,Sanur
4,673,1,13,Seminyak Beach,Seminyak
5,737,1,76,Tanah Lot Temple,"Kecamatan Kediri, Kabupaten Tabanan, Beraban 8..."
6,748,1,72,Tegalalang Rice Terrace,"Jalan Raya Ceking, Tegalalang 80517 Indonesia"
7,749,1,93,Tegenungan Waterfall,"Jl. Raya Tegenungan, Kemenuh, Ubud 80581 Indon..."
8,824,1,76,Uluwatu Temple,"Jl. Raya Uluwatu Southern part of Bali, Pecatu..."
9,841,1,92,Waterbom Bali,"Jl. Kartika Plaza, Kuta 80361 Indonesia"



Rows from Updated_Item used by Transaction:


,AttractionId,AttractionCityId,AttractionTypeId,Attraction,AttractionAddress
0,369,1,13,Kuta Beach - Bali,Kuta
1,481,1,13,Nusa Dua Beach,"Semenanjung Nusa Dua, Nusa Dua 80517 Indonesia"
2,640,1,63,Sacred Monkey Forest Sanctuary,"Jl. Monkey Forest, Ubud 80571 Indonesia"
3,650,1,13,Sanur Beach,Sanur
4,673,1,13,Seminyak Beach,Seminyak
5,737,1,76,Tanah Lot Temple,"Kecamatan Kediri, Kabupaten Tabanan, Beraban 8..."
6,748,1,72,Tegalalang Rice Terrace,"Jalan Raya Ceking, Tegalalang 80517 Indonesia"
7,749,1,93,Tegenungan Waterfall,"Jl. Raya Tegenungan, Kemenuh, Ubud 80581 Indon..."
8,824,1,76,Uluwatu Temple,"Jl. Raya Uluwatu Southern part of Bali, Pecatu..."
9,841,1,92,Waterbom Bali,"Jl. Kartika Plaza, Kuta 80361 Indonesia"


In [21]:
# ============================================================
# STEP 3B: CHECK VISIT MODE
# ============================================================

print("VisitMode values in Transaction:")
print(sorted(transaction["VisitMode"].unique()))

print("\nVisitModeId values in Mode:")
print(sorted(mode["VisitModeId"].unique()))

print("\nMode lookup table:")
display(mode)

VisitMode values in Transaction:
[np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5)]

VisitModeId values in Mode:
[np.int64(0), np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5)]

Mode lookup table:


,VisitModeId,VisitMode
0,0,-
1,1,Business
2,2,Couples
3,3,Family
4,4,Friends
5,5,Solo


#### Check the actual categorical values

In [22]:
# ============================================================
# STEP 3C: CHECK CATEGORICAL VALUES
# ============================================================

categorical_columns = {
    "Country": ["Country"],
    "Region": ["Region"],
    "Continent": ["Continent"],
    "City": ["CityName"],
    "Type": ["AttractionType"],
    "Mode": ["VisitMode"],
    "Item": ["Attraction"]
}

for table_name, columns in categorical_columns.items():
    print("=" * 70)
    print(f"{table_name}")
    print("=" * 70)

    df = tables[table_name]

    for column in columns:
        print(f"\n{column}:")
        print("Unique values:", df[column].nunique(dropna=False))
        print(df[column].dropna().sort_values().head(30).tolist())

Country

Country:
Unique values: 164
['-', 'Afghanistan', 'Albania', 'Algeria', 'Angola', 'Antigua and Barbuda', 'Argentina', 'Armenia', 'Asia', 'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh', 'Barbados', 'Belarus', 'Belgium', 'Belize', 'Bhutan', 'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil', 'Brunei Darussalam', 'Bulgaria', 'Burkina Faso', 'Cambodia', 'Cameroon', 'Canada']
Region

Region:
Unique values: 22
['-', 'Australia', 'Caribbean', 'Central Africa', 'Central America', 'Central Asia', 'Central Europe', 'East Africa', 'East Asia', 'Eastern Europe', 'Middle East', 'North Africa', 'Northern America', 'Northern Europe', 'Oceania', 'South America', 'South Asia', 'South East Asia', 'Southern Africa', 'Southern Europe', 'West Africa', 'Western Europe']
Continent

Continent:
Unique values: 6
['-', 'Africa', 'America', 'Asia', 'Australia & Oceania', 'Europe']
City

CityName:
Unique values: 8768
['-', '-', '-', '-', '-', '-', '-', '-', 'Aachen', 'Aalbo

What we've established
 Transaction → User

All 33,530 transaction users exist in User.

 Transaction → Attraction

All 30 attraction IDs exist in both Item and Updated_Item.

More importantly, the 30 transaction attractions are identical in both tables. Therefore, for this project we can use Item.xlsx as the lookup table because it is the smaller, relevant table and contains every attraction used by the transactions.

Updated_Item.xlsx contains additional attractions that aren't referenced by our transactions. We will retain it as a source/audit table but won't unnecessarily merge 1,698 rows into the final transaction dataset.

 User → City

All non-null user CityIds exist in City.

 City → Country → Region → Continent

All foreign-key relationships are valid.

 Missing CityIds

Four users have missing CityId. We will not guess these values.

 Missing CityName

CityId = 6879 has a missing CityName. We'll investigate it before deciding whether it can be recovered.

 VisitMode

Transaction.VisitMode contains 1–5, while Mode contains 0–5.

## Final data-quality investigation

In [23]:
# ============================================================
# STEP 3: FINAL DATA-QUALITY INVESTIGATION
# ============================================================

# ------------------------------------------------------------
# 1. Investigate the city with missing CityName
# ------------------------------------------------------------

print("=" * 70)
print("1. CITY WITH MISSING CITY NAME")
print("=" * 70)

missing_city_name = city[city["CityName"].isnull()]

display(missing_city_name)

print("\nCountry information:")
display(
    country[
        country["CountryId"].isin(missing_city_name["CountryId"])
    ]
)


# ------------------------------------------------------------
# 2. Check whether CityId 6879 is referenced by users
# ------------------------------------------------------------

print("\n" + "=" * 70)
print("2. REFERENCES TO CITY ID 6879")
print("=" * 70)

users_in_missing_city = user[user["CityId"] == 6879]

print("Number of users assigned to CityId 6879:",
      len(users_in_missing_city))

display(users_in_missing_city.head(20))


# ------------------------------------------------------------
# 3. Check the four users with missing CityId
# ------------------------------------------------------------

print("\n" + "=" * 70)
print("3. USERS WITH MISSING CITY ID")
print("=" * 70)

display(
    user[user["CityId"].isnull()]
)


# ------------------------------------------------------------
# 4. Check whether their countries have a likely city
# ------------------------------------------------------------

print("\n" + "=" * 70)
print("4. CITIES AVAILABLE FOR THEIR COUNTRIES")
print("=" * 70)

affected_country_ids = (
    user[user["CityId"].isnull()]["CountryId"]
    .dropna()
    .unique()
)

print("Affected CountryIds:", affected_country_ids)

for country_id in affected_country_ids:

    print("\nCountryId:", country_id)

    country_info = country[
        country["CountryId"] == country_id
    ]

    display(country_info)

    cities_for_country = city[
        city["CountryId"] == country_id
    ]

    print("Number of cities:", len(cities_for_country))

    display(cities_for_country.head(20))


# ------------------------------------------------------------
# 5. Validate VisitMode relationship
# ------------------------------------------------------------

print("\n" + "=" * 70)
print("5. VISIT MODE VALIDATION")
print("=" * 70)

transaction_modes = set(transaction["VisitMode"])
lookup_modes = set(mode["VisitModeId"])

print("Transaction VisitMode values:",
      sorted(transaction_modes))

print("Mode lookup VisitModeId values:",
      sorted(lookup_modes))

print(
    "Transaction VisitMode values missing from Mode:",
    transaction_modes - lookup_modes
)

print("\nTransaction VisitMode frequency:")
print(transaction["VisitMode"].value_counts().sort_index())

print("\nMode lookup:")
display(mode)


# ------------------------------------------------------------
# 6. Validate AttractionTypeId relationship
# ------------------------------------------------------------

print("\n" + "=" * 70)
print("6. ATTRACTION TYPE VALIDATION")
print("=" * 70)

item_type_ids = set(item["AttractionTypeId"])
type_ids = set(type_["AttractionTypeId"])

print("Item AttractionTypeId values:",
      sorted(item_type_ids))

print("Type AttractionTypeId values:",
      sorted(type_ids))

print(
    "Item AttractionTypeIds missing from Type:",
    item_type_ids - type_ids
)


# ------------------------------------------------------------
# 7. Check Rating values
# ------------------------------------------------------------

print("\n" + "=" * 70)
print("7. RATING VALIDATION")
print("=" * 70)

print("Unique Rating values:")
print(sorted(transaction["Rating"].unique()))

print("\nRating frequency:")
print(transaction["Rating"].value_counts().sort_index())

print("\nMinimum Rating:",
      transaction["Rating"].min())

print("Maximum Rating:",
      transaction["Rating"].max())


# ------------------------------------------------------------
# 8. Check VisitYear and VisitMonth
# ------------------------------------------------------------

print("\n" + "=" * 70)
print("8. DATE VALIDATION")
print("=" * 70)

print("VisitYear values:")
print(sorted(transaction["VisitYear"].unique()))

print("\nVisitMonth values:")
print(sorted(transaction["VisitMonth"].unique()))

print("\nInvalid VisitMonth values:")
print(
    transaction[
        ~transaction["VisitMonth"].between(1, 12)
    ][["VisitYear", "VisitMonth"]]
)

1. CITY WITH MISSING CITY NAME


,CityId,CityName,CountryId
6879,6879,NaN,151



Country information:


,CountryId,Country,RegionId
151,151,Italy,20



2. REFERENCES TO CITY ID 6879
Number of users assigned to CityId 6879: 0


,UserId,ContinentId,RegionId,CountryId,CityId



3. USERS WITH MISSING CITY ID


,UserId,ContinentId,RegionId,CountryId,CityId
2279,7175,5,17,135,NaN
6027,17595,1,4,22,NaN
21303,56972,5,17,135,NaN
25494,67461,5,17,135,NaN



4. CITIES AVAILABLE FOR THEIR COUNTRIES
Affected CountryIds: [135  22]

CountryId: 135


,CountryId,Country,RegionId
135,135,Switzerland,17


Number of cities: 123


,CityId,CityName,CountryId
6205,6205,Aarau,135
6206,6206,Aclens,135
6207,6207,Affoltern am Albis,135
6208,6208,Altdorf,135
6209,6209,Asuel,135
6210,6210,Baar,135
6211,6211,Baden,135
6212,6212,Bannwil,135
6213,6213,Basel,135
6214,6214,Bassersdorf,135



CountryId: 22


,CountryId,Country,RegionId
22,22,South Africa,4


Number of cities: 120


,CityId,CityName,CountryId
93,93,Alberton,22
94,94,Ballito,22
95,95,Barrydale,22
96,96,Bedfordview,22
97,97,Benoni,22
98,98,Bergville,22
99,99,Bloemfontein,22
100,100,Bloubergstrand,22
101,101,Bluff,22
102,102,Boksburg,22



5. VISIT MODE VALIDATION
Transaction VisitMode values: [1, 2, 3, 4, 5]
Mode lookup VisitModeId values: [0, 1, 2, 3, 4, 5]
Transaction VisitMode values missing from Mode: set()

Transaction VisitMode frequency:
VisitMode
1      623
2    21620
3    15217
4    10945
5     4525
Name: count, dtype: int64

Mode lookup:


,VisitModeId,VisitMode
0,0,-
1,1,Business
2,2,Couples
3,3,Family
4,4,Friends
5,5,Solo



6. ATTRACTION TYPE VALIDATION
Item AttractionTypeId values: [2, 10, 13, 19, 34, 44, 45, 61, 63, 64, 72, 76, 82, 84, 91, 92, 93]
Type AttractionTypeId values: [2, 10, 13, 19, 34, 44, 45, 61, 63, 64, 72, 76, 82, 84, 91, 92, 93]
Item AttractionTypeIds missing from Type: set()

7. RATING VALIDATION
Unique Rating values:
[np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5)]

Rating frequency:
Rating
1     1263
2     2035
3     7730
4    17966
5    23936
Name: count, dtype: int64

Minimum Rating: 1
Maximum Rating: 5

8. DATE VALIDATION
VisitYear values:
[np.int64(2013), np.int64(2014), np.int64(2015), np.int64(2016), np.int64(2017), np.int64(2018), np.int64(2019), np.int64(2020), np.int64(2021), np.int64(2022)]

VisitMonth values:
[np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8), np.int64(9), np.int64(10), np.int64(11), np.int64(12)]

Invalid VisitMonth values:
Empty DataFrame
Columns: [VisitYear, VisitMonth]
Index: []


-What we can conclude

Rating: completely valid — only 1–5. No outliers/invalid values.

VisitYear: valid — 2013–2022.

VisitMonth: valid — 1–12.

VisitMode: every transaction value maps correctly to Mode.xlsx.

AttractionTypeId: every Item value maps correctly to Type.xlsx.

CityId 6879: has no city name, but no user references it, so it won't affect the transaction dataset.

4 users have missing CityId: there is no reliable way to determine their cities. We should leave them missing rather than guess.

Important: those users' country/region/continent information is still valid and can be retained.

Updated_Item.xlsx does not need to be merged because the transaction-relevant attraction records matched Item.xlsx.

## Build the final cleaned dataset

In [24]:
# ============================================================
# STEP 4: CLEAN, MERGE AND CREATE FINAL DATASET
# ============================================================

import pandas as pd
import numpy as np
from pathlib import Path

print("=" * 70)
print("STARTING FINAL DATA CLEANING AND MERGING")
print("=" * 70)

# ------------------------------------------------------------
# 1. Make copies so original raw tables remain unchanged
# ------------------------------------------------------------

tx = transaction.copy()
usr = user.copy()
cty = city.copy()
cntry = country.copy()
reg = region.copy()
cont = continent.copy()
itm = item.copy()
typ = type_.copy()
md = mode.copy()

print("Original transaction rows:", len(tx))


# ------------------------------------------------------------
# 2. Standardize text fields
# ------------------------------------------------------------

text_columns = {
    "Country": cntry,
    "Region": reg,
    "Continent": cont,
    "CityName": cty,
    "Attraction": itm,
    "AttractionAddress": itm,
    "AttractionType": typ,
    "VisitMode": md
}

for column_name, df in text_columns.items():
    if column_name in df.columns:
        df[column_name] = (
            df[column_name]
            .astype("string")
            .str.strip()
        )

print("\nText fields standardized:")
print("- Leading/trailing whitespace removed")
print("- Original capitalization preserved")


# ------------------------------------------------------------
# 3. Convert placeholder "-" to missing values
# ------------------------------------------------------------

for df in [cntry, reg, cont, cty, typ, md]:
    for col in df.columns:
        if df[col].dtype == "string":
            df[col] = df[col].replace("-", pd.NA)

print("\nPlaceholder '-' converted to missing values where applicable.")


# ------------------------------------------------------------
# 4. Validate lookup-table keys before merging
# ------------------------------------------------------------

key_checks = {
    "User.UserId": (usr, "UserId"),
    "City.CityId": (cty, "CityId"),
    "Country.CountryId": (cntry, "CountryId"),
    "Region.RegionId": (reg, "RegionId"),
    "Continent.ContinentId": (cont, "ContinentId"),
    "Item.AttractionId": (itm, "AttractionId"),
    "Type.AttractionTypeId": (typ, "AttractionTypeId"),
    "Mode.VisitModeId": (md, "VisitModeId")
}

print("\n" + "=" * 70)
print("PRIMARY KEY VALIDATION")
print("=" * 70)

for name, (df, key) in key_checks.items():
    duplicate_keys = df[key].duplicated().sum()
    print(f"{name}: {duplicate_keys} duplicate keys")

    if duplicate_keys > 0:
        raise ValueError(
            f"{name} contains duplicate keys. "
            "Fix this before merging to avoid multiplying transaction rows."
        )

print("All lookup keys are unique.")


# ------------------------------------------------------------
# 5. Rename Transaction.VisitMode to VisitModeId
# ------------------------------------------------------------

tx = tx.rename(columns={"VisitMode": "VisitModeId"})


# ------------------------------------------------------------
# 6. Create VisitDate
# ------------------------------------------------------------

tx["VisitDate"] = pd.to_datetime(
    {
        "year": tx["VisitYear"],
        "month": tx["VisitMonth"],
        "day": 1
    },
    errors="coerce"
)

invalid_dates = tx["VisitDate"].isna().sum()

print("\n" + "=" * 70)
print("DATE CREATION")
print("=" * 70)
print("Invalid VisitDate values:", invalid_dates)

if invalid_dates > 0:
    raise ValueError("Some VisitYear/VisitMonth combinations produced invalid dates.")

print("VisitDate created successfully.")


# ------------------------------------------------------------
# 7. Merge User information
# ------------------------------------------------------------

user_columns = [
    "UserId",
    "ContinentId",
    "RegionId",
    "CountryId",
    "CityId"
]

final_df = tx.merge(
    usr[user_columns],
    on="UserId",
    how="left",
    validate="many_to_one"
)

print("\nAfter User merge:", final_df.shape)


# ------------------------------------------------------------
# 8. Merge City information
# ------------------------------------------------------------

city_lookup = cty[
    ["CityId", "CityName"]
].copy()

final_df = final_df.merge(
    city_lookup,
    on="CityId",
    how="left",
    validate="many_to_one"
)

print("After City merge:", final_df.shape)


# ------------------------------------------------------------
# 9. Merge Country information
# ------------------------------------------------------------

country_lookup = cntry[
    ["CountryId", "Country"]
].copy()

final_df = final_df.merge(
    country_lookup,
    on="CountryId",
    how="left",
    validate="many_to_one"
)

print("After Country merge:", final_df.shape)


# ------------------------------------------------------------
# 10. Merge Region information
# ------------------------------------------------------------

region_lookup = reg[
    ["RegionId", "Region"]
].copy()

final_df = final_df.merge(
    region_lookup,
    on="RegionId",
    how="left",
    validate="many_to_one"
)

print("After Region merge:", final_df.shape)


# ------------------------------------------------------------
# 11. Merge Continent information
# ------------------------------------------------------------

continent_lookup = cont[
    ["ContinentId", "Continent"]
].copy()

final_df = final_df.merge(
    continent_lookup,
    on="ContinentId",
    how="left",
    validate="many_to_one"
)

print("After Continent merge:", final_df.shape)


# ------------------------------------------------------------
# 12. Merge Item information
# ------------------------------------------------------------

item_columns = [
    "AttractionId",
    "AttractionCityId",
    "AttractionTypeId",
    "Attraction",
    "AttractionAddress"
]

final_df = final_df.merge(
    itm[item_columns],
    on="AttractionId",
    how="left",
    validate="many_to_one"
)

print("After Item merge:", final_df.shape)


# ------------------------------------------------------------
# 13. Merge Attraction Type
# ------------------------------------------------------------

type_lookup = typ[
    ["AttractionTypeId", "AttractionType"]
].copy()

final_df = final_df.merge(
    type_lookup,
    on="AttractionTypeId",
    how="left",
    validate="many_to_one"
)

print("After Attraction Type merge:", final_df.shape)


# ------------------------------------------------------------
# 14. Merge Visit Mode description
# ------------------------------------------------------------

mode_lookup = md[
    ["VisitModeId", "VisitMode"]
].copy()

final_df = final_df.merge(
    mode_lookup,
    on="VisitModeId",
    how="left",
    validate="many_to_one"
)

print("After Visit Mode merge:", final_df.shape)


# ------------------------------------------------------------
# 15. Reorder columns
# ------------------------------------------------------------

desired_columns = [
    "TransactionId",
    "UserId",

    "VisitDate",
    "VisitYear",
    "VisitMonth",

    "VisitModeId",
    "VisitMode",

    "AttractionId",
    "Attraction",
    "AttractionAddress",
    "AttractionCityId",
    "AttractionTypeId",
    "AttractionType",

    "Rating",

    "CityId",
    "CityName",
    "CountryId",
    "Country",
    "RegionId",
    "Region",
    "ContinentId",
    "Continent"
]

# Keep only columns that exist
desired_columns = [
    col for col in desired_columns
    if col in final_df.columns
]

final_df = final_df[desired_columns]


# ------------------------------------------------------------
# 16. Final data types
# ------------------------------------------------------------

final_df["VisitDate"] = pd.to_datetime(final_df["VisitDate"])

# Keep IDs as nullable integers because CityId contains missing values
for col in [
    "UserId",
    "VisitModeId",
    "AttractionId",
    "AttractionCityId",
    "AttractionTypeId",
    "CityId",
    "CountryId",
    "RegionId",
    "ContinentId"
]:
    if col in final_df.columns:
        final_df[col] = final_df[col].astype("Int64")


# ------------------------------------------------------------
# 17. Final validation
# ------------------------------------------------------------

print("\n" + "=" * 70)
print("FINAL VALIDATION")
print("=" * 70)

print("Final dataset shape:", final_df.shape)

print("\nOriginal transaction rows:", len(transaction))
print("Final dataset rows:", len(final_df))

if len(final_df) != len(transaction):
    raise ValueError(
        "Row count changed! Check the merge logic before saving."
    )

print("✓ Transaction row count preserved.")


# Check duplicate transaction IDs
duplicate_transactions = final_df["TransactionId"].duplicated().sum()

print("\nDuplicate TransactionIds:", duplicate_transactions)

if duplicate_transactions > 0:
    raise ValueError("Duplicate TransactionIds detected.")

print("✓ TransactionId remains unique.")


# Check missing important descriptive fields
print("\nMissing values in important columns:")

important_columns = [
    "TransactionId",
    "UserId",
    "VisitDate",
    "VisitMode",
    "Attraction",
    "AttractionType",
    "Rating",
    "Country",
    "Region",
    "Continent"
]

for col in important_columns:
    print(f"{col}: {final_df[col].isna().sum()}")


# Rating validation
invalid_ratings = final_df[
    ~final_df["Rating"].between(1, 5)
]

print("\nInvalid ratings:", len(invalid_ratings))

if len(invalid_ratings) > 0:
    raise ValueError("Invalid ratings detected.")


# Date validation
print(
    "VisitDate range:",
    final_df["VisitDate"].min(),
    "to",
    final_df["VisitDate"].max()
)


# ------------------------------------------------------------
# 18. Preview
# ------------------------------------------------------------

print("\n" + "=" * 70)
print("FINAL DATASET PREVIEW")
print("=" * 70)

display(final_df.head(10))

print("\nFinal columns:")
print(final_df.columns.tolist())

print("\nFinal dataset info:")
final_df.info()

STARTING FINAL DATA CLEANING AND MERGING
Original transaction rows: 52930

Text fields standardized:
- Leading/trailing whitespace removed
- Original capitalization preserved

Placeholder '-' converted to missing values where applicable.

PRIMARY KEY VALIDATION
User.UserId: 0 duplicate keys
City.CityId: 0 duplicate keys
Country.CountryId: 0 duplicate keys
Region.RegionId: 0 duplicate keys
Continent.ContinentId: 0 duplicate keys
Item.AttractionId: 0 duplicate keys
Type.AttractionTypeId: 0 duplicate keys
Mode.VisitModeId: 0 duplicate keys
All lookup keys are unique.

DATE CREATION
Invalid VisitDate values: 0
VisitDate created successfully.

After User merge: (52930, 12)
After City merge: (52930, 13)
After Country merge: (52930, 14)
After Region merge: (52930, 15)
After Continent merge: (52930, 16)
After Item merge: (52930, 20)
After Attraction Type merge: (52930, 21)
After Visit Mode merge: (52930, 22)

FINAL VALIDATION
Final dataset shape: (52930, 22)

Original transaction rows: 52930
F

,TransactionId,UserId,VisitDate,VisitYear,VisitMonth,VisitModeId,VisitMode,AttractionId,Attraction,AttractionAddress,...,AttractionType,Rating,CityId,CityName,CountryId,Country,RegionId,Region,ContinentId,Continent
0,3,70456,2022-10-01,2022,10,2,Couples,640,Sacred Monkey Forest Sanctuary,"Jl. Monkey Forest, Ubud 80571 Indonesia",...,Nature & Wildlife Areas,5,4341,Guildford,163,United Kingdom,21,Western Europe,5,Europe
1,8,7567,2022-10-01,2022,10,4,Friends,640,Sacred Monkey Forest Sanctuary,"Jl. Monkey Forest, Ubud 80571 Indonesia",...,Nature & Wildlife Areas,5,464,Ontario,48,Canada,8,Northern America,2,America
2,9,79069,2022-10-01,2022,10,3,Family,640,Sacred Monkey Forest Sanctuary,"Jl. Monkey Forest, Ubud 80571 Indonesia",...,Nature & Wildlife Areas,5,774,Brazil,54,Brazil,9,South America,2,America
3,10,31019,2022-10-01,2022,10,3,Family,640,Sacred Monkey Forest Sanctuary,"Jl. Monkey Forest, Ubud 80571 Indonesia",...,Nature & Wildlife Areas,3,583,Zurich,135,Switzerland,17,Central Europe,5,Europe
4,15,43611,2022-10-01,2022,10,2,Couples,640,Sacred Monkey Forest Sanctuary,"Jl. Monkey Forest, Ubud 80571 Indonesia",...,Nature & Wildlife Areas,3,1396,Manchester,163,United Kingdom,21,Western Europe,5,Europe
5,18,43471,2022-10-01,2022,10,4,Friends,640,Sacred Monkey Forest Sanctuary,"Jl. Monkey Forest, Ubud 80571 Indonesia",...,Nature & Wildlife Areas,5,1806,San Francisco,51,United States,8,Northern America,2,America
6,22,76492,2022-09-01,2022,9,2,Couples,640,Sacred Monkey Forest Sanctuary,"Jl. Monkey Forest, Ubud 80571 Indonesia",...,Nature & Wildlife Areas,4,31,Arusha,9,Tanzania,2,East Africa,1,Africa
7,23,20977,2022-10-01,2022,10,4,Friends,640,Sacred Monkey Forest Sanctuary,"Jl. Monkey Forest, Ubud 80571 Indonesia",...,Nature & Wildlife Areas,5,8258,Edinburgh,163,United Kingdom,21,Western Europe,5,Europe
8,28,18655,2022-09-01,2022,9,2,Couples,640,Sacred Monkey Forest Sanctuary,"Jl. Monkey Forest, Ubud 80571 Indonesia",...,Nature & Wildlife Areas,3,114,Durban,22,South Africa,4,Southern Africa,1,Africa
9,29,62907,2022-09-01,2022,9,2,Couples,640,Sacred Monkey Forest Sanctuary,"Jl. Monkey Forest, Ubud 80571 Indonesia",...,Nature & Wildlife Areas,5,50,Alexandria,51,United States,8,Northern America,2,America



Final columns:
['TransactionId', 'UserId', 'VisitDate', 'VisitYear', 'VisitMonth', 'VisitModeId', 'VisitMode', 'AttractionId', 'Attraction', 'AttractionAddress', 'AttractionCityId', 'AttractionTypeId', 'AttractionType', 'Rating', 'CityId', 'CityName', 'CountryId', 'Country', 'RegionId', 'Region', 'ContinentId', 'Continent']

Final dataset info:
<class 'pandas.DataFrame'>
RangeIndex: 52930 entries, 0 to 52929
Data columns (total 22 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   TransactionId      52930 non-null  int64         
 1   UserId             52930 non-null  Int64         
 2   VisitDate          52930 non-null  datetime64[us]
 3   VisitYear          52930 non-null  int64         
 4   VisitMonth         52930 non-null  int64         
 5   VisitModeId        52930 non-null  Int64         
 6   VisitMode          52930 non-null  string        
 7   AttractionId       52930 non-null  Int64         
 8  

What this cell is doing

The important thing is that we're using left joins throughout.

That means:

Every one of the original 52,930 transactions must remain.

We're not deleting transactions just because some user/city information is missing.

For the four users without CityId, their transactions will remain, but CityId and CityName will be missing. Their Country, Region and Continent are still retained from the User table.

### Save cleaned_data.csv

In [26]:
# ============================================================
# STEP 6: SAVE FINAL CLEANED DATASET
# ============================================================

from pathlib import Path

output_path = Path("../data/processed/cleaned_data.csv")

final_df.to_csv(
    output_path,
    index=False
)

print("=" * 70)
print("CLEANED DATASET SAVED")
print("=" * 70)

print("File:", output_path.resolve())
print("Rows:", len(final_df))
print("Columns:", len(final_df.columns))
print("File size:", round(output_path.stat().st_size / (1024 * 1024), 2), "MB")

CLEANED DATASET SAVED
File: /home/kusuma-rani-sutar/tourism-analytics/data/processed/cleaned_data.csv
Rows: 52930
Columns: 22
File size: 9.25 MB
